### PR TITLE
Adding ClientEncryptionS3PinotFS

### PR DESCRIPTION
**Pull request by xiangfu0**
https://docs.amazonaws.cn/en_us/sdk-for-java/v1/developer-guide/examples-crypto-masterkey.html

To use client side encryption, you can specify either 
* `kmsCmkId` which contains the id of the KMS key as the value
* `aesHexSecret` which contains a custom generated AES 256 key.

The Filesystem for scheme `s3` also needs to be configured to `org.apache.pinot.plugin.filesystem.ClientEncryptionS3PinotFS`

### Controller conf for deep storage
```
controller.data.dir=s3://my-bucket/quickstart-data
controller.local.temp.dir=/tmp/pinot/data/controller/index

pinot.controller.storage.factory.class.s3=org.apache.pinot.plugin.filesystem.ClientEncryptionS3PinotFS
pinot.controller.segment.fetcher.protocols=file,http,s3
pinot.controller.segment.fetcher.s3.class=org.apache.pinot.common.utils.fetcher.PinotFSSegmentFetcher
pinot.controller.storage.factory.s3.region=us-east-1
pinot.controller.storage.factory.s3.accessKey=XXXX
pinot.controller.storage.factory.s3.secretKey=XXXX
pinot.controller.storage.factory.s3.kmsCmkId=703373a8-3e4b-4eb8-8a3f-03699a0d0927
```

### Ingestion spec
```
pinotFSSpecs:
    - scheme: s3
      className: org.apache.pinot.plugin.filesystem.S3PinotFS
      configs:
        region: 'us-east-1'
        accessKey: 'XXXX'
        secretKey: 'XXXX'
        kmsCmkId: '703373a8-3e4b-4eb8-8a3f-03699a0d0927'
```

### CODE REVIEW

File: pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/ClientEncryptionS3PinotFS.java [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+    return _s3Client.getObject(getObjectRequest).getObjectContent();
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    super.close();
+  }
+
+  public static void main(String[] args)
```
Comment (KKcorps) [inline pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/ClientEncryptionS3PinotFS.java:573]: We should remove this

File: pinot-plugins/pom.xml [hunk 2]
```diff
@@ -150,6 +162,10 @@
                       <pattern>org.apache.avro</pattern>
                       <shadedPattern>${shade.prefix}.org.apache.avro</shadedPattern>
                     </relocation>
+                    <relocation>
+                      <pattern>org.apache.avro</pattern>
+                      <shadedPattern>${shade.prefix}.org.apache.avro</shadedPattern>
```
Comment (KKcorps) [inline pinot-plugins/pom.xml:167]: seems duplicate of previous line

File: pinot-plugins/pom.xml [hunk 3]
```diff
@@ -90,7 +90,19 @@
         <plugins>
           <plugin>
             <artifactId>maven-shade-plugin</artifactId>
-            <version>3.2.1</version>
+            <version>3.3.0</version>
```
Comment (KKcorps) [inline pinot-plugins/pom.xml:93]: reason for upgrading this?

**Comment by KKcorps** [comment]
Weird that AWS Java SDK 2 doesn't have encryption based S3 client. The issue seems to be open from last 5 years - https://github.com/aws/aws-sdk-java-v2/issues/34

File: pinot-plugins/pom.xml (hunk 3 above)
Comment (xiangfu0) [reply pinot-plugins/pom.xml:93]: cause this bouncycastle lib has multi-release, so the shade plugin doesn't have a compilation error. The fix is to add dependency of asm for this. I can revert the version upgrade and retry.

**Comment by xiangfu0** [comment]
> Weird that AWS Java SDK 2 doesn't have encryption based S3 client. The issue seems to be open from last 5 years - [aws/aws-sdk-java-v2#34](https://github.com/aws/aws-sdk-java-v2/issues/34)

Right, so for the client-side encryption, we need to use new lib and write a new PinotFS 🤦

**Review by KKcorps — APPROVED** [review]

**Comment by KKcorps** [comment]
Added some unit tests as well.

**Comment by xiangfu0** [comment]
Close since no further actions.

**Comment by rino-kadijk** [comment]
I recommend also looking into a key rotation strategy

**Comment by vmarchaud** [comment]
> Close since no further actions.

Not sure to understand why it didn't got merged ? The S3 plugin doesn't support [AES 256](https://github.com/apache/pinot/blob/master/pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/S3PinotFS.java#L202) encryption currently right ?

**Pull request closed without merging**